docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md

```diff
@@ -0,0 +1,366 @@
+
+    // Core API
+    virtual int set(const char *key, const void *buffer, size_t size, uint32_t create_flags);
+    virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size, offset = 0);
+    virtual int get_info(const char *key, info_t *info);
+    virtual int remove(const char *key);
+ 
+    // Incremental set API
+    virtual int set_start(set_handle_t *handle, const char *key, size_t final_data_size, uint32_t create_flags);
+    virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);
+    virtual int set_finalize(set_handle_t handle);
+ 
+    // Key iterator
+    virtual int iterator_open(iterator_t *it, const char *pattern=NULL);
+    virtual int iterator_next(iterator_t it, char *key, size_t key_size, size_t *actual_key_size);
+    virtual int iterator_close(iterator_t it);
+    
+private:
+    Mutex _mutex;
+    size_t *_max_keys;
+    size_t *_num_keys;
+    FileSystem *_fs;
+    bool _is_initialized;
+}
+```
+
+### Important data structures
+
+```C++
+// Key metadata
+typedef struct {
+    uint32_t magic;
+    uint16_t metadata_size;
+    uint16_t revision;
+    uint32_t ext_flags;
+    uint32_t data_size;
+} key_metadata_t;
+
+// incremental set handle
+typedef struct {
+    char *key;
+    uint32_t create_flags;
+} inc_set_handle_t;
+
+// iterator handle
+typedef struct {
+    void *dir_handle;
+    char *pattern;
+} key_iterator_handle_t;
+```
+
+
+### Initialization and reset
+
+**init function**
+
+Header:  
+`virtual int init();`
+
+Pseudo code:  
+- if `_is_initialized` return OK
+- Create and take `_mutex`
+- Create the FileSystemStore directory if not existing
+- Using the FileSystem `dir_size` API, count number of files and store in `_num_keys` 
+- Set `_is_initialized` to true
+- Release `_mutex`     
+  
+**deinit function**
+
+Header:  
+`virtual int deinit();`
+
+Pseudo code:  
+- if not `_is_initialized` return OK
+- Take `_mutex` 
+- Set `_is_initialized` to false
+- Release `_mutex`   
+
+**reset function**
+
+Header:  
+`virtual int reset();`
+
+Pseudo code:  
+- Take `_mutex`
+- Delete all files under the FileSystemStore directory 
+- Set `_num_keys` to 0
+- Release `_mutex`   
+
+### Core APIs
+
+**set function**
+
+Header:  
+`virtual int set(const char *key, const void *buffer, size_t size, uint32_t create_flags);`
+
+Pseudo code:  
+- if not `_is_initialized` return error
+- Call `set_start` with all fields and a local `set_handle_t` variable
+- Call `set_add_data` with `buffer` and `size`
+- Call `set_finalize`
+- Return OK
+
+**get function**
+
+Header:  
+`virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size, offset = 0);`
+
+Pseudo code:  
+- if not `_is_initialized` return error
+- Find file `key` under the FileSystemStore directory. If not existing, return not found error
+- Take `_mutex` 
+- Using the FileSystem `stat` API, extract file size, set `actual_size` as file size - metadata size
```

Comment:
Earlier it said that 'get' is implemented "using open-read-close"
Do we want to mention it here?